### PR TITLE
Parser: more unfinished type recovery

### PR DESCRIPTION
**Pull request by auduchinok**
Continues https://github.com/dotnet/fsharp/pull/15410 and adds recovery for a missed case:
```fsharp
type T() =
```

### CODE REVIEW

**Comment by auduchinok** [comment]
Please don't merge it yet, I'll continue some work here tomorrow.

**Comment by auduchinok** [comment]
How much do we care about parser recovery for missing `=` in the following case? It doesn't seem like a popular mistake people make to me.

```fsharp
type MissingEquals A of int
```

There was [a special case](https://github.com/dotnet/fsharp/blob/main/src/Compiler/pars.fsy#L1591) added some time ago, and it makes it very tricky to have any progress here. I've spent many hours today, trying to find workaround it, but for now my conclusion is it doesn't work well with precedence of `typ` rule that is used inside type abbreviation declarations:

```fsharp
type Abbr = int
```

Changing that precedence rule seems too dangerous now, because `typ` is used everywhere in the grammar. Besides that, there may be other affecting precedences specified somewhere in the grammar, that we'd uncover later.

If the special case for missing `=` is removed, the parser starts to produce the usual syntax error about the unexpected token. That seems sufficient for the Visual Studio quick fix. The downside is subsequent nodes aren't parsed properly. The good thing is that the much more common unfinished `type T() =` is parsed properly and subsequent code is not broken during typing.

**Comment by smoothdeveloper** [comment]
> It doesn't seem like a popular mistake people make to me.

For DU & enums, it is not frequent, but when switching from other languages to F#, I sometimes miss it on records and classes.

But since the error message is giving what the end user needs:

> Unexpected token in type definition. Expected '=' after the type 'A'.

I feel that the recovery for this particular case doesn't make it a big issue, especially if it comes with a code fix.

I'm wondering also, if FCS wouldn't benefit from a using stale type checking results, whenever the syntax tree comes broken, if there was a way to split the "dirty" segment and recompose the parts?

**Comment by auduchinok** [comment]
> For DU & enums, it is not frequent, but when switching from other languages to F#, I sometimes miss it on records and classes.

My guess is they would also type `{` in, and I doubt that the parser is ready for that. Or is there some popular syntax where neither `=` nor `{` is required?

**Comment by T-Gro** [comment]
> 

I think it is a good tradeoff to bring in your change as it is.
Unfinished typing is a daily scenario for every user.

**Comment by auduchinok** [comment]
> I think it is a good tradeoff to bring in your change as it is.

Thanks for your reply! Yes, I've dropped that rule, and it simplified things considerably. It has also allowed improving errors and ranges in other places.

**Comment by auduchinok** [comment]
This is ready. @psfinaki Thanks a lot for helping with the change!

File: src/Compiler/pars.fsy [hunk 1]
```diff
@@ -1588,24 +1588,6 @@ tyconDefn:
...
            SynTypeDefn($1, SynTypeDefnRepr.Simple(SynTypeDefnSimpleRepr.None($1.Range), $1.Range), [], None, $1.Range, trivia) }
 
-  | typeNameInfo opt_equals tyconDefnRhsBlock
-     { match $2 with
-       | Some _ -> ()
-       | None ->
-            let (SynComponentInfo(_, _, _, lid, _, _, _, _)) = $1
-            // While the spec doesn't allow long idents here, the parser doesn't enforce this, so take one ident
-            let typeNameId = List.last lid
-            raiseParseErrorAt (rhs parseState 2) (FSComp.SR.parsEqualsMissingInTypeDefinition(typeNameId.ToString()))
```
Comment (psfinaki) [inline src/Compiler/pars.fsy:1598]: So I guess this error should be removed now, otherwise - let's get this in :)

**Review by psfinaki — APPROVED** [review]

**Pull request merged**